sdk/storage/azure_storage_blob/src/partitioned_transfer/download.rs

```diff
@@ -0,0 +1,314 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+use std::{collections::VecDeque, ops::Range, task::Poll};
+
+use azure_core::http::AsyncRawResponse;
+use bytes::Bytes;
+
+use crate::models::content_range::ContentRange;
+
+use super::*;
+
+#[async_trait::async_trait]
+pub(crate) trait PartitionedDownloadBehavior {
+    async fn transfer_range(&self, range: Range<u64>) -> AzureResult<AsyncRawResponse>;
+}
+
+/// Holds either a future to be polled or its output to be persisted.
+/// Allows tracking pending futures and their completed results in the
+/// same data structure.
+enum PollPersist<Out, Fut> {
+    Ready(Out),
+    Pending(Fut),
+}
+type OpsFuture = Pin<Box<dyn Future<Output = Result<Bytes, azure_core::Error>>>>;
+
+/// Returns a stream that runs up to parallel-many ranged downloads at a time.
+///
+/// Downloads are stored in-order. The returned stream will produce an item only when the next
+/// download in the sequence has been buffered, regardless of the state of any other downloads.
+/// This means completed ranged downloads may sit for a while while earlier ones complete.
+///
+/// A download that has completed buffering but has not yet returned its buffer in the resulting
+/// stream will still count when determining current running operations. This is so the stream can
+/// promise its buffered bytes do not exceed parallel * partition_size.
+pub(crate) async fn download<'a, T: PartitionedDownloadBehavior>(
+    parallel: usize,
+    partition_size: usize,
```

Comment:
In upload, these are both type `NonZero`.  I am assuming you would want the same here? Tracing through the code it seems like you would get unintended effects if either are 0:

- If `parallel=0`, L58 and follow codeblock will never execute, so it will never saturate the jobs of the rest of the ranges
- if `partition_size=0`, L42 will be `client.transfer_range(0..0).await?;`, not really sure what happens there 😄 but doesn't seem good